lib/fastlane/plugin/firebase_app_distribution/helper/firebase_app_distribution_auth_client.rb

```diff
@@ -4,56 +4,87 @@ module Fastlane
+      FIREBASE_TOOLS_DESC = "the firebase login:ci command"
 
-      def fetch_auth_token(google_service_path)
-        if !google_service_path.nil? && !google_service_path.empty?
-          service_account(google_service_path)
-        elsif ENV["FIREBASE_TOKEN"]
-          firebase_token(ENV["FIREBASE_TOKEN"])
-        elsif ENV["GOOGLE_APPLICATION_CREDENTIALS"]
-          service_account(ENV["GOOGLE_APPLICATION_CREDENTIALS"])
-        elsif (refresh_token = refresh_token_from_firebase_tools)
-          firebase_token(refresh_token)
-        else
-          UI.crash!(ErrorMessage::MISSING_CREDENTIALS)
+      # Returns the auth token for any of the auth methods (Firebase CLI token,
+      # Google service account, firebase-tools). To ensure that a specific
+      # auth method is used, unset all other auth variables/parameters to nil/empty
+      #
+      # args
+      #   google_service_path - Absolute path to the Google service account file
+      #   firebase_cli_token - Firebase CLI refresh token from login action or
+      #                        CI environment
+      #
+      # env variables
+      #   GOOGLE_APPLICATION_CREDENTIALS - see google_service_path
+      #   FIREBASE_TOKEN - see firebase_cli_token
+      #
+      # Crashes if given invalid or missing credentials
+      def fetch_auth_token(google_service_path, firebase_cli_token)
+        begin
+          if !google_service_path.nil? && !google_service_path.empty?
+            auth_method = SERVICE_ACCOUNT_PARAM_DESC
+            token = service_account(google_service_path)
+          elsif !firebase_cli_token.nil? && !firebase_cli_token.empty?
+            auth_method = FIREBASE_TOKEN_PARAM_DESC
+            token = firebase_token(firebase_cli_token)
+          elsif !ENV["FIREBASE_TOKEN"].nil? && !ENV["FIREBASE_TOKEN"].empty?
+            auth_method = FIREBASE_TOKEN_ENV_DESC
+            token = firebase_token(ENV["FIREBASE_TOKEN"])
+          elsif !ENV["GOOGLE_APPLICATION_CREDENTIALS"].nil? && !ENV["GOOGLE_APPLICATION_CREDENTIALS"].empty?
+            auth_method = SERVICE_ACCOUNT_ENV_DESC
+            token = service_account(ENV["GOOGLE_APPLICATION_CREDENTIALS"])
+          elsif (refresh_token = refresh_token_from_firebase_tools)
+            auth_method = FIREBASE_TOOLS_DESC
+            token = firebase_token(refresh_token)
+          else
+            UI.user_error!(ErrorMessage::MISSING_CREDENTIALS)
+          end
+        rescue => error
+          unless auth_method.nil? # nil only when no credentials are passed in
+            UI.error("App Distribution encountered an error while authenticating with #{auth_method}")
+          end
+          UI.user_error!(error.message)
         end
+        UI.success("Authenticated with #{auth_method}")
+        token
       end
 
       private
 
       def refresh_token_from_firebase_tools
-        if ENV["XDG_CONFIG_HOME"]
+        if !ENV["XDG_CONFIG_HOME"].nil? && !ENV["XDG_CONFIG_HOME"].empty?
```

Comment:
The `!` here make this hard to read. We can get rid of them by changing the order of the if else i.e.

```
if ENV["XDG_CONFIG_HOME"].nil? || ENV["XDG_CONFIG_HOME"].empty?
  config_path = File.expand_path(".config/configstore/firebase-tools.json", "~")
else
 config_path = File.expand_path("configstore/firebase-tools.json", ENV["XDG_CONFIG_HOME"]) 
end
```

Reply:
good catch!